### PR TITLE
[Merged by Bors] - feat: Hermite-Minkowski theorem

### PR DESCRIPTION
**Pull request by xroblot**
We prove  lower bounds for the abs. value of the discriminant of a number field in terms of its degree and deduce from it Hermite-Minkowski theorem: a nontrivial number field has nontrivial absolute discriminant. 

```lean
theorem abs_discr_ge (h : 1 < finrank ℚ K) :
    (4 / 9 : ℝ) * (3 * π / 4) ^ finrank ℚ K ≤ |discr K| 

theorem discr_gt_one (h : 1 < finrank ℚ K) : 2 < |discr K|
```

---

- [x] depends on: #8030 
- [x] depends on: #8481 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/NumberField/Discriminant.lean [hunk 1]
```diff
@@ -80,6 +90,108 @@ theorem _root_.NumberField.mixedEmbedding.volume_fundamentalDomain_latticeBasis
...
+      congr
+      rw [← card_add_two_mul_card_eq_rank, Nat.cast_add, Nat.cast_mul, Nat.cast_ofNat]
+      ring
+    _ = (4 / π) ^ NrComplexPlaces K * (finrank ℚ K).factorial * Real.sqrt |discr K| := by
+      rw [show ‖discr K‖ = |(discr K : ℝ)| by rfl, zpow_mul, show (2:ℝ) ^ (2:ℤ) = 4 by norm_cast,
+        div_pow, inv_eq_one_div, div_pow, one_pow, zpow_coe_nat]
+      ring
+
+theorem abs_discr_ge (h : 1 < finrank ℚ K) :
+    ((4:ℝ) / (9:ℝ)) * (3 * π / 4) ^ finrank ℚ K ≤ |discr K| := by
```
Comment (alreadydone) [inline Mathlib/NumberTheory/NumberField/Discriminant.lean:141]:
```suggestion
    (4 / 9 : ℝ) * (3 * π / 4) ^ finrank ℚ K ≤ |discr K| := by
```
(untested)

**Review by alreadydone — COMMENTED** [review]
Wonderful!

File: Mathlib/NumberTheory/NumberField/Discriminant.lean (hunk 1 above)
Comment (xroblot) [reply Mathlib/NumberTheory/NumberField/Discriminant.lean]: Yes, that works, thanks! I tried some other combinations that resulted in having some ↑ here and there, but this works fine.

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding.lean [hunk 2]
```diff
@@ -525,50 +532,199 @@ section convexBodySum
...
+
+theorem convexBodySumFactor_ne_top : convexBodySumFactor K ≠ ⊤ := by
+  refine mul_ne_top (mul_ne_top (pow_ne_top two_ne_top) ?_) ?_
+  · rw [show (2:ℝ≥0∞) = (2:NNReal) by rfl, ← ENNReal.coe_div two_ne_zero]
+    exact pow_ne_top coe_ne_top
+  · exact inv_ne_top.mpr <| Nat.cast_ne_zero.mpr (Nat.factorial_ne_zero _)
+
+open MeasureTheory MeasureTheory.Measure Real
+
+theorem convexBodySum_volume :
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/CanonicalEmbedding.lean:677]:
```suggestion
open MeasureTheory MeasureTheory.Measure Real in
theorem convexBodySum_volume :
```
If those are used only in this theorem.

**Review by riccardobrasca — APPROVED** [review]
LGTM, thanks!

I am just a little bit worried about the various `open Classical` (I fear they can create diamonds if `K = ℚ`). I guess that everything comes from `Mathlib.NumberTheory.NumberField.Embeddings`. Tagging @eric-wieser as an expert in this kind of stuff.

**Comment by eric-wieser** [comment]
> LGTM, thanks!
> 
> I am just a little bit worried about the various `open Classical` (I fear they can create diamonds if `K = ℚ`). I guess that everything comes from `Mathlib.NumberTheory.NumberField.Embeddings`. Tagging @eric-wieser as an expert in this kind of stuff.

Could you point out where you think such a diamond might appear? I'm happy to investigate, but it would be easier for me if I knew where things looked questionable to you!

**Comment by eric-wieser** [comment]
I had a quick go at removing `Classical` from this file; it seems this would require adding
```
variable [DecidablePred (IsReal (K := K))] [DecidablePred (IsComplex (K := K))]
```
and `DecidableEq (InfinitePlace K)` in a few places; though it's not obvious to me whether that makes any sense.

**Comment by riccardobrasca** [comment]
> I had a quick go at removing `Classical` from this file; it seems this would require adding
> 
> ```
> variable [DecidablePred (IsReal (K := K))] [DecidablePred (IsComplex (K := K))]
> ```
> 
> and `DecidableEq (InfinitePlace K)` in a few places; though it's not obvious to me whether that makes any sense.

OK, these will probably not make any sense without computable reals, so we are safe.

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding.lean [hunk 3]
```diff
@@ -524,50 +531,198 @@ section convexBodySum
...
-  refine ⟨fun hx => ?_, fun h => h.elim⟩
-  · rw [Set.mem_setOf] at hx
-    have : 0 ≤ ∑ w, ‖x.1 w‖ + 2 * ∑ w, ‖x.2 w‖ := by
-      refine add_nonneg ?_ ?_
-      · exact Finset.sum_nonneg (fun _ _ => norm_nonneg _)
-      · exact mul_nonneg zero_le_two (Finset.sum_nonneg (fun _ _ => norm_nonneg _))
-    linarith
+abbrev convexBodySum : Set (E K)  := { x | convexBodySumFun x ≤ B }
+
+theorem convexBodySum_volume_eq_zero {B} (hB : B ≤ 0) : volume (convexBodySum K B) = 0 := by
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/CanonicalEmbedding.lean:613]:
```suggestion
theorem convexBodySum_volume_eq_zero_of_le_zero {B} (hB : B ≤ 0) : volume (convexBodySum K B) = 0 := by
```

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding.lean [hunk 4]
```diff
@@ -524,50 +531,198 @@ section convexBodySum
...
     mixedEmbedding K x ∈ (convexBodySum K B) ↔
       ∑ w : InfinitePlace K, (mult w) * w.val x ≤ B := by
-  simp_rw [Set.mem_setOf_eq, mixedEmbedding, RingHom.prod_apply, Pi.ringHom_apply,
+  simp_rw [Set.mem_setOf_eq, mixedEmbedding, RingHom.prod_apply, convexBodySumFun, Pi.ringHom_apply,
     ← Complex.norm_real, embedding_of_isReal_apply, norm_embedding_eq, mult, Nat.cast_ite, ite_mul,
     Finset.sum_ite, Finset.filter_congr (fun _ _ => not_isReal_iff_isComplex), Finset.mul_sum,
     ← Finset.sum_subtype_eq_sum_filter, Finset.subtype_univ, Nat.cast_one, one_mul, Nat.cast_ofNat]
   rfl
 
 theorem convexBodySum_symmetric (x : E K) (hx : x ∈ (convexBodySum K B)) :
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/CanonicalEmbedding.lean:634]:
```suggestion
theorem convexBodySum_symmetric {x : E K} (hx : x ∈ (convexBodySum K B)) :
```

File: Mathlib/NumberTheory/NumberField/Discriminant.lean [hunk 5]
```diff
@@ -94,6 +103,107 @@ theorem _root_.NumberField.mixedEmbedding.volume_fundamentalDomain_latticeBasis
...
+          π⁻¹ ^ (NrComplexPlaces K) := by
+      congr
+      rw [← card_add_two_mul_card_eq_rank, Nat.cast_add, Nat.cast_mul, Nat.cast_ofNat]
+      ring
+    _ = (4 / π) ^ NrComplexPlaces K * (finrank ℚ K).factorial * Real.sqrt |discr K| := by
+      rw [show ‖discr K‖ = |(discr K : ℝ)| by rfl, zpow_mul, show (2:ℝ) ^ (2:ℤ) = 4 by norm_cast,
+        div_pow, inv_eq_one_div, div_pow, one_pow, zpow_coe_nat]
+      ring
+
+theorem abs_discr_ge (h : 1 < finrank ℚ K) :
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/Discriminant.lean:152]: Here `K` can be implicit, but as you prefer.

File: Mathlib/NumberTheory/NumberField/Discriminant.lean [hunk 6]
```diff
@@ -94,6 +103,107 @@ theorem _root_.NumberField.mixedEmbedding.volume_fundamentalDomain_latticeBasis
...
+        · simp_rw [add_mul, one_mul, pow_succ, Nat.factorial_succ]
+          field_simp; ring
+        · rw [_root_.le_div_iff (by positivity), pow_succ]
+          convert (mul_le_mul h_m this (by positivity) (by positivity)) using 1
+          field_simp; ring
+      refine le_trans (le_of_eq (by field_simp; norm_num)) (one_add_mul_le_pow ?_ (2 * m))
+      exact le_trans (by norm_num : (-2:ℝ) ≤ 0) (by positivity)
+
+/-- **Hermite-Minkowski Theorem**. A nontrivial number field has nontrivial discriminant. -/
+theorem discr_gt_one (h : 1 < finrank ℚ K) : 2 < |discr K| := by
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/Discriminant.lean:191]: Same here.

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors d+

**Pull request closed without merging**